crates/sui-core/src/epoch/data_removal.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use mysten_metrics::spawn_monitored_task;
+use narwhal_config::Epoch;
+use std::fs;
+use std::path::PathBuf;
+use tokio::sync::mpsc;
+
+pub struct EpochDataRemover {
+    base_path: PathBuf,
+    tx_remove: mpsc::Sender<Epoch>,
+}
+
+impl EpochDataRemover {
+    pub fn new(base_path: PathBuf) -> Self {
+        let (tx_remove, _rx_remove) = mpsc::channel(1);
+        Self {
+            base_path,
+            tx_remove,
+        }
+    }
+
+    pub async fn run(&mut self) {
+        let (tx_remove, mut rx_remove) = mpsc::channel(1);
+        self.tx_remove = tx_remove;
+        let base_path = self.base_path.clone();
+        spawn_monitored_task!(async {
+            while let Some(epoch) = rx_remove.recv().await {
+                remove_old_epoch_data(base_path.clone(), epoch);
+            }
+        });
+    }
+
+    pub async fn remove_old_data(&self, latest_closed_epoch: Epoch) {
+        let result = self.tx_remove.send(latest_closed_epoch).await;
+        if result.is_err() {
+            tracing::error!(
+                "Error sending message to data removal task for epoch {:?}",
+                latest_closed_epoch,
```

Comment:
If we hit this error because the previous call to remove the epoch data is still running, then we will do nothing and this will be called at the next epoch change as an automatic "retry". The cleanup task then removes all the old epochs, so it will include the one we missed. So if it was a slow as a one-off we don't need to do anything special.

Reply:
I believe that if an error is returned while we try to send to the channel that would be due to the [Receiver being dropped/closed](https://doc.rust-lang.org/std/sync/mpsc/struct.Sender.html#method.send) - so eventually this will not be received. I think this is strong indication that the delete task is not running and we do have two options:
1. panic 
2. we just record the error

we can just go with 2 for now as you have here - but we shouldn't expect that the loop is still running